brion/plugin/compartmentReportHDF5.cpp

```diff
@@ -637,16 +757,123 @@ void CompartmentReportHDF5::_allocateDataSet()
+
+    HighFive::DataSetAccessProps caching;
+    size_t blocksPerFrame =
+        (_targetMapping.frameSize + chunkDims[1] - 1) / chunkDims[1];
+    caching.add(
+        HighFive::Caching(blocksPerFrame * _blockSize, blocksPerFrame + 1));
 
     _data.reset(new HighFive::DataSet(
-        _file->createDataSet<float>("data", dataspace, chunking)));
+        _file->createDataSet<float>("data", dataspace, chunking, caching)));
 
     detail::addStringAttribute(*_data, "dunit", _dunit);
 }
+
+void CompartmentReportHDF5::_parseWriteOptions(const URI& uri)
+{
+    for (auto i = uri.queryBegin(); i != uri.queryEnd(); ++i)
+    {
+        const auto& key = i->first;
+        const auto& value = i->second;
+        if (key == "cells_to_frames")
+        {
+            if (value == "inf")
+            {
+                _cellsToFramesRatio = std::numeric_limits<float>::infinity();
+            }
+            else
+            {
+                std::string::size_type idx;
+                const float ratio = std::stof(value, &idx);
+                if (idx != value.size())
+                {
+                    std::cerr << "Warning: invalid value for "
+                                 "cell_to_frames H5 report parameter"
+                              << std::endl;
+                }
+                else
+                {
+                    _cellsToFramesRatio = ratio;
+                }
+            }
+        }
+    }
+}
+
+void CompartmentReportHDF5::_reopenDataSet(size_t cacheSizeHint)
+{
+    // Getting the chunk dims
+    auto properties = H5Dget_create_plist(_data->getId());
+    hsize_t chunkDims[2];
+    H5Pget_chunk(properties, 2, chunkDims);
+    H5Pclose(properties);
+    // And the frame count
+    const size_t frameCount = _data->getSpace().getDimensions()[0];
+
+    if (chunkDims[0] == 0)
+        return; // Nothing to configure for this case.
+
+    // Need to close the dataset first. Otherwise internal H5 reference
+    // couning prevents the caching parameters from taking effect.
+    _data.reset();
+
+    if (cacheSizeHint == 0)
+    {
+        // The default cache configuration from HDF5 gives very bad performance
+        // because it can only hold 3 blocks and it causes block eviction and
+        // re-reads all the time. The preferred behaviour if not hint is given
+        // is to disable the cache.
+        HighFive::DataSetAccessProps accessProps;
+        accessProps.add(HighFive::Caching(0, 0));
+        _data.reset(
+            new HighFive::DataSet(_file->getDataSet("data", accessProps)));
+        return;
+    }
+
+    // We pick the number of slots so it can at least hold a full trace or
+    // a full frame. Note that the maximum cache size allowed by this number
+    // and the actual maximum cache size are independent.
+    // To avoid collisions reading traces we also need a number of slots s,
+    // such as blocks[1] ∤ s. At the same time we want blocks[0] ∤ s.
```

Comment:
Write "blocks[1] does not divide by s" to avoid using utf-8 symbols.

Reply:
Is it a concern having utf-8 symbols in 2018?